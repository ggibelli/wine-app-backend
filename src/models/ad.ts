/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
