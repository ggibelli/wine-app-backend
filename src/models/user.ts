/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
