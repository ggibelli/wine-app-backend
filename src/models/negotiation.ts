import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { IAdDoc } from './ad';
import { IMessageDoc } from './message';
import { TypeAd } from '../types';
import mongooseUniqueValidator from 'mongoose-unique-validator';

export interface INegotiation {
  createdBy: IUserDoc['_id'];
  ad: IAdDoc['_id'];
  forUserAd: IUserDoc['_id'];
  messages?: mongoose.Types.Array<IMessageDoc['_id']>;
  isConcluded: boolean;
  dateCreated: Date;
  type: TypeAd;
}

export interface NegotiationGraphQl extends INegotiation {
  _id: mongoose.Types.ObjectId;
}

export interface INegotiationDoc extends INegotiation, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const negotiationSchemaFields: Record<keyof INegotiation, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
    required: true,
  },
  ad: {
    type: Schema.Types.ObjectId,
    ref: 'Ad',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  isConcluded: {
    type: Boolean,
    default: false,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

const negotiationSchema = new Schema(negotiationSchemaFields);

negotiationSchema.index({ createdBy: 1, ad: 1 }, { unique: true });
negotiationSchema.plugin(mongooseUniqueValidator);
export const Negotiation = mongoose.model<INegotiationDoc>(
  'Negotiation',
  negotiationSchema
);
