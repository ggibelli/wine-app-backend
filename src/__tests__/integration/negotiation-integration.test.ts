/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
//import { ObjectId } from 'mongodb';
import { User } from '../../models/user';
import { Ad, AdGraphQl } from '../../models/ad';
import { users, ads } from '../../tests/mocksTests';
import { NegotiationInput } from '../../generated/graphql';

import { Negotiation, NegotiationGraphQl } from '../../models/negotiation';
//import { ObjectId } from 'mongodb';
import cron from 'cron';

const fakeStart = jest.fn(() => fakeStop());
const fakeStop = jest.fn(() => null);
const FakeCron = jest.fn(() => ({
  stop: fakeStop,
  start: fakeStart,
  fireOnTick: fakeStart,
}));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
cron.CronJob = FakeCron;

const { query, mutate, setOptions } = testClient;

const NEGOTIATIONS_USER = gql`
  query GetNegotiationsUser($user: ID!) {
    negotiationsWithUser(forUserAd: $user) {
      createdBy {
        firstName
      }
      ad {
        content
      }
      forUserAd {
        firstName
      }
      isConcluded
    }
  }
`;

const NEGOTIATIONS_AD = gql`
  query GetNegotiationsAd($ad: ID!) {
    negotiationsForAd(ad: $ad) {
      createdBy {
        firstName
      }
      ad {
        content
      }
      forUserAd {
        firstName
      }
      isConcluded
    }
  }
`;

const NEGOTIATIONS = gql`
  {
    negotiations {
      negotiations {
        createdBy {
          firstName
        }
        ad {
          content
        }
        forUserAd {
          firstName
        }
        isConcluded
      }
    }
  }
`;

const NEGOTIATION = gql`
  query GetNegotiation($id: ID!) {
    negotiation(id: $id) {
      createdBy {
        firstName
      }
      ad {
        content
        ... on AdWine {
          wineName
        }
        ... on AdGrape {
          vineyardName
        }
      }
      forUserAd {
        firstName
      }
      isConcluded
    }
  }
`;

const CREATE_NEGOTIATION = gql`
  mutation CreateNegotiation($negotiation: NegotiationInput!) {
    createNegotiation(negotiation: $negotiation) {
      response {
        createdBy {
          firstName
        }
        ad {
          ... on AdWine {
            wineName
          }
          ... on AdGrape {
            vineyardName
          }
        }
        forUserAd {
          firstName
        }
        isConcluded
      }
      errors {
        name
        text
      }
    }
  }
`;

const UPDATE_NEGOTIATION = gql`
  mutation UpdateNegotiation($negotiation: NegotiationInputUpdate!) {
    updateNegotiation(negotiation: $negotiation) {
      response {
        createdBy {
          firstName
        }
        ad {
          ... on AdWine {
            wineName
          }
          ... on AdGrape {
            vineyardName
          }
        }
        forUserAd {
          firstName
        }
        isConcluded
      }
      errors {
        name
        text
      }
    }
  }
`;

const DELETE_NEGOTIATION = gql`
  mutation DeleteNegotiation($id: ID!) {
    deleteNegotiation(id: $id) {
      response {
        createdBy {
          firstName
        }
        ad {
          ... on AdWine {
            wineName
          }
          ... on AdGrape {
            vineyardName
          }
        }
        forUserAd {
          firstName
        }
        isConcluded
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_NOT_VERIFIED = gql`
  mutation {
    login(email: "prova@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const adsMock = ads();
  const user = new User(usersMock[0]);
  const otherUser = new User(usersMock[1]);
  const thirdUser = new User(usersMock[2]);
  const ad = new Ad({ ...adsMock[0], postedBy: user });
  const otherAd = new Ad({ ...adsMock[1], postedBy: otherUser });
  //const thirdAd = new Ad({ ...adsMock[2], postedBy: thirdUser });
  const negotiation = new Negotiation({
    createdBy: otherUser,
    ad: ad,
    forUserAd: ad.postedBy,
    type: ad.typeAd,
  });
  const otherNegotiation = new Negotiation({
    createdBy: thirdUser,
    ad: otherAd,
    forUserAd: otherAd.postedBy,
    type: otherAd.typeAd,
  });
  await user.save();
  await otherUser.save();
  await thirdUser.save();
  await ad.save();
  await otherAd.save();
  await negotiation.save();
  await otherNegotiation.save();
});

describe('Integration test negotiations', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('query negotiations fails if not logged in', async () => {
    const res = await query(NEGOTIATIONS);
    expect(res).toMatchSnapshot();
  });

  it('query negotiationsWithUser fails if not logged in', async () => {
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await query(NEGOTIATIONS_USER, {
      variables: { user: ad[0].postedBy.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query negotiationsForAd fails if not logged in', async () => {
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await query(NEGOTIATIONS_AD, {
      variables: { ad: ad[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query single negotiation fails if not logged', async () => {
    const negotiation: NegotiationGraphQl[] = await Negotiation.find({})
      .lean()
      .exec();
    const res = await query(NEGOTIATION, {
      variables: { id: negotiation[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create negotiation mutation fails if not logged in', async () => {
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const negotiation: NegotiationInput = {
      ad: ad[0]._id.toString(),
      forUserAd: ad[0].postedBy.toString(),
      type: ad[0].typeAd,
    };
    const res = await mutate(CREATE_NEGOTIATION, {
      variables: { negotiation: negotiation },
    });
    expect(res).toMatchSnapshot();
  });

  it('create negotiation mutation fails if user not verified', async () => {
    const data: any = await mutate(LOGIN_VALID_NOT_VERIFIED);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const negotiation: NegotiationInput = {
      ad: ad[1]._id.toString(),
      forUserAd: ad[1].postedBy.toString(),
      type: ad[1].typeAd,
    };
    const res = await mutate(CREATE_NEGOTIATION, {
      variables: { negotiation },
    });
    expect(res).toMatchSnapshot();
  });

  it('create negotiation mutation successfull', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const negotiation: NegotiationInput = {
      ad: ad[1]._id.toString(),
      forUserAd: ad[1].postedBy.toString(),
      type: ad[1].typeAd,
    };
    const res = await mutate(CREATE_NEGOTIATION, {
      variables: { negotiation },
    });
    expect(fakeStart).toBeCalledTimes(1);
    expect(fakeStop).toBeCalledTimes(1);
    expect(res).toMatchSnapshot();
  });

  it('create negotiation mutation fails if is same ad and user', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const negotiation: NegotiationInput = {
      ad: ad[1]._id.toString(),
      forUserAd: ad[1].postedBy.toString(),
      type: ad[1].typeAd,
    };
    const res: any = await mutate(CREATE_NEGOTIATION, {
      variables: { negotiation },
    });
    expect(res.data.createNegotiation.response).toBeNull();
    expect(res.data.createNegotiation.errors).toHaveLength(1);
  });

  it('query negotiations succeds if logged in and admin, all are shown', async () => {
    const res = await query(NEGOTIATIONS);
    expect(res).toMatchSnapshot();
  });

  it('query negotiationsWithUser succeds if logged in', async () => {
    const ad = await Ad.find({}).lean().exec();
    const res = await query(NEGOTIATIONS_USER, {
      variables: { user: ad[1].postedBy.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query negotiationsForAd succeds if logged in', async () => {
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await query(NEGOTIATIONS_AD, {
      variables: { ad: ad[1]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query single negotiation succeds if logged', async () => {
    const negotiation: NegotiationGraphQl[] = await Negotiation.find({})
      .lean()
      .exec();
    const res = await query(NEGOTIATION, {
      variables: { id: negotiation[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('update negotiation mutation succeds if logged in and same user', async () => {
    const user = await User.findOne({ firstName: 'Giovanni' });
    const negotiationToUpdate: NegotiationGraphQl | null = await Negotiation.findOne(
      {
        createdBy: user,
      }
    )
      .lean()
      .exec();
    const negotiation = {
      _id: negotiationToUpdate?._id.toString(),
      isConcluded: true,
    };
    const res = await mutate(UPDATE_NEGOTIATION, {
      variables: { negotiation: negotiation },
    });
    expect(res).toMatchSnapshot();
  });

  it('update negotiation mutation fails if logged in and not same user', async () => {
    const user = await User.findOne({ firstName: 'Luigetto' });
    const negotiationToUpdate: NegotiationGraphQl | null = await Negotiation.findOne(
      {
        createdBy: user,
      }
    )
      .lean()
      .exec();
    const negotiation = {
      _id: negotiationToUpdate?._id.toString(),
      isConcluded: true,
    };
    const res = await mutate(UPDATE_NEGOTIATION, {
      variables: { negotiation },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete negotiation mutation succeds if logged in and same user', async () => {
    const user = await User.findOne({ firstName: 'Giovanni' });
    const negotiationToDelete: NegotiationGraphQl | null = await Negotiation.findOne(
      {
        createdBy: user,
      }
    )
      .lean()
      .exec();
    const res = await mutate(DELETE_NEGOTIATION, {
      variables: { id: negotiationToDelete?._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete negotiation mutation fails if logged in and not same user', async () => {
    const user = await User.findOne({ firstName: 'Mariuccio' });
    const negotiationToDelete: NegotiationGraphQl | null = await Negotiation.findOne(
      {
        createdBy: user,
      }
    )
      .lean()
      .exec();
    const res = await mutate(DELETE_NEGOTIATION, {
      variables: { id: negotiationToDelete?._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query negotiations succeds if logged in and not admin, only user negotiation shown', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(NEGOTIATIONS);
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
