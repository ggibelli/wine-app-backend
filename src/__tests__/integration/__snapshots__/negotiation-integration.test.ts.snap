// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test negotiations create negotiation mutation fails if not logged in 1`] = `
Object {
  "data": Object {
    "createNegotiation": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "createNegotiation",
      ],
    },
  ],
}
`;

exports[`Integration test negotiations create negotiation mutation fails if user not verified 1`] = `
Object {
  "data": Object {
    "createNegotiation": Object {
      "errors": Array [
        Object {
          "name": "AuthorizationError",
          "text": "You need to verify your account",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test negotiations create negotiation mutation successfull 1`] = `
Object {
  "data": Object {
    "createNegotiation": Object {
      "errors": Array [],
      "response": Object {
        "ad": Object {
          "wineName": "Vino vendo",
        },
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
    },
  },
}
`;

exports[`Integration test negotiations delete negotiation mutation fails if logged in and not same user 1`] = `
Object {
  "data": Object {
    "deleteNegotiation": Object {
      "errors": Array [
        Object {
          "name": "General error",
          "text": "Errors during the negotiation delete, only creator can delete negotiation",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test negotiations delete negotiation mutation succeds if logged in and same user 1`] = `
Object {
  "data": Object {
    "deleteNegotiation": Object {
      "errors": Array [],
      "response": Object {
        "ad": Object {
          "wineName": "Vino vendo",
        },
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": true,
      },
    },
  },
}
`;

exports[`Integration test negotiations query negotiations fails if not logged in 1`] = `
Object {
  "data": Object {
    "negotiations": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "negotiations",
      ],
    },
  ],
}
`;

exports[`Integration test negotiations query negotiations succeds if logged in and admin, all are shown 1`] = `
Object {
  "data": Object {
    "negotiations": Array [
      Object {
        "ad": Object {
          "content": "wow",
        },
        "createdBy": Object {
          "firstName": "Mariuccio",
        },
        "forUserAd": Object {
          "firstName": "Giovanni",
        },
        "isConcluded": false,
      },
      Object {
        "ad": Object {
          "content": "waaow",
        },
        "createdBy": Object {
          "firstName": "Luigetto",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
      Object {
        "ad": Object {
          "content": "waaow",
        },
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
    ],
  },
}
`;

exports[`Integration test negotiations query negotiations succeds if logged in and not admin, only user negotiation shown 1`] = `
Object {
  "data": Object {
    "negotiations": Array [
      Object {
        "ad": Object {
          "content": "wow",
        },
        "createdBy": Object {
          "firstName": "Mariuccio",
        },
        "forUserAd": Object {
          "firstName": "Giovanni",
        },
        "isConcluded": false,
      },
      Object {
        "ad": Object {
          "content": "waaow",
        },
        "createdBy": Object {
          "firstName": "Luigetto",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
    ],
  },
}
`;

exports[`Integration test negotiations query negotiationsForAd fails if not logged in 1`] = `
Object {
  "data": Object {
    "negotiationsForAd": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "negotiationsForAd",
      ],
    },
  ],
}
`;

exports[`Integration test negotiations query negotiationsForAd succeds if logged in 1`] = `
Object {
  "data": Object {
    "negotiationsForAd": Array [
      Object {
        "ad": Object {
          "content": "waaow",
        },
        "createdBy": Object {
          "firstName": "Luigetto",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
      Object {
        "ad": Object {
          "content": "waaow",
        },
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
    ],
  },
}
`;

exports[`Integration test negotiations query negotiationsWithUser fails if not logged in 1`] = `
Object {
  "data": Object {
    "negotiationsWithUser": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "negotiationsWithUser",
      ],
    },
  ],
}
`;

exports[`Integration test negotiations query negotiationsWithUser succeds if logged in 1`] = `
Object {
  "data": Object {
    "negotiationsWithUser": Array [
      Object {
        "ad": Object {
          "content": "waaow",
        },
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": false,
      },
    ],
  },
}
`;

exports[`Integration test negotiations query single negotiation fails if not logged 1`] = `
Object {
  "data": Object {
    "negotiation": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "negotiation",
      ],
    },
  ],
}
`;

exports[`Integration test negotiations query single negotiation succeds if logged 1`] = `
Object {
  "data": Object {
    "negotiation": Object {
      "ad": Object {
        "content": "wow",
        "wineName": "Vino",
      },
      "createdBy": Object {
        "firstName": "Mariuccio",
      },
      "forUserAd": Object {
        "firstName": "Giovanni",
      },
      "isConcluded": false,
    },
  },
}
`;

exports[`Integration test negotiations update negotiation mutation fails if logged in and not same user 1`] = `
Object {
  "data": Object {
    "updateNegotiation": Object {
      "errors": Array [
        Object {
          "name": "General error",
          "text": "Errors during the negotiation update, only creator can update negotiation",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test negotiations update negotiation mutation succeds if logged in and same user 1`] = `
Object {
  "data": Object {
    "updateNegotiation": Object {
      "errors": Array [],
      "response": Object {
        "ad": Object {
          "wineName": "Vino vendo",
        },
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUserAd": Object {
          "firstName": "Mariuccio",
        },
        "isConcluded": true,
      },
    },
  },
}
`;
